### ISSUE TITLE
Pauli Trotter Evolution fails to exponentiate the YYYY operator

### ISSUE THREAD

**Issue by shifubear** [Qiskit/qiskit#9298]
### Environment

- **Qiskit Terra version**: 0.21.2
- **Python version**: 3.10.8
- **Operating system**: macOS Ventura 13.0.1

### What is happening?

The [`PauliTrotterEvolution`](https://qiskit.org/documentation/stubs/qiskit.opflow.evolutions.PauliTrotterEvolution.html) class fails to exponentiate any Pauli operator that contains a $YYYY$ component. Attempting to do so gives the error message 
"ValueError: Operator contains complex coefficients, which are not supported." 

This causes issues for chemistry applications, for example when running phase estimation for Hamiltonians on 4 qubits (and likely any multiple of 4), since there will likely be a component that is all $Y$s. 

### How can we reproduce the issue?

The following code is a minimal working example. If reproduced correctly, the code should fail with the error "ValueError: Operator contains complex coefficients, which are not supported."

```
from qiskit.opflow import I, Y, PauliTrotterEvolution

y4_hamiltonian = (
  (I ^ I ^ I ^ I)
  + (Y ^ Y ^ Y ^ Y)
).to_pauli_op()

evolution = PauliTrotterEvolution()
exp_ham = evolution.convert(y4_hamiltonian.exp_i())
```

### What should happen?

Any Pauli operator should be able to be exponentiated without issue. This is a very critical bug for any chemistry applications. 

### Cause 
I found out the cause of the bug, which I will do my best to explain briefly here. If there are any questions please let me know and I will clarify any points. 

The key equation is the line 132 in the [`sparse_pauli_op.py`](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py) file, which I have copied here for ease of explanation:
```
self._coeffs = np.asarray((-1j) ** (phase - count_y) * coeffs, dtype=complex)
```

In the above example, `phase = np.array([0])`, `count_y = np.array([4])`, and `coeffs = np.array([1.+0.j])`. 
The following snippet allows reproducing the above situation: 
```
import numpy as np 

phase = np.array([0])
count_y = np.array([4])
coeffs = np.array([1.+0.j])
result = np.asarray((-1j) ** (phase - count_y) * coeffs, dtype=complex)
print(result)
```
The above code will `1.+0.j`, which is what we want. 

However, it turns out that both the `phase` and `count_y` variables are initialized as numpy arrays with `dtype=uint8`. This forces the subtraction to equal `phase - count_y = [252]`. This is because the variables are initialized to the smallest possible `dtype`s that can contain the variable (See line 691 of [`base_pauli.py`](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/symplectic/base_pauli.py)). 
Due to floating point error, this causes the code above to return `[1.-5.8857324e-15j]`, which is a "complex coefficient" that isn't allowed by the `PauliTrotterEvolution` class. This situation can be reproduced with the following code snippet: 
```
import numpy as np

phase = np.array([0], dtype=np.uint8)
count_y = np.array([4], dtype=np.uint8)
coeffs = np.array([1.+0.j])
result = np.asarray((-1j) ** (phase - count_y) * coeffs, dtype=complex)
print(result)
```

### Suggestions for Fixes 
I'm not sure what the most "Qiskit-like" fix would be, but I have two suggestions that would avoid this from happening. 
1. Prevent casting to 'unsigned' data types in the [`base_pauli.py`](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/symplectic/base_pauli.py) class. This will make sure that the subtraction is done correctly, and not lead to large exponents that can generate floating point errors like above. 
2. Take mod 4 of `phase - count_y` in [`sparse_pauli_op.py`](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py) so that the exponent is always in [0, 1, 2, 3]. Since we are exponentiating $-i$ here, mathematically taking mod 4 will give the correct answer. 

I think suggestion 2 is a better fix, but since the root cause is related to suggestion 1, I wonder if fixing that will help avoid similar errors in other parts of the code. If a Qiskit developer can shed light on which fix is better for the project I'm happy to make the fix.

**Comment by woodsp-ibm** [comment]
@ikkoham While the title seemed to imply an opflow issue it seems the author has narrowed down the issue to what seems to be due to numpy types used in quantum info SparsePauliOp and BasePauli. Hence I labelled the issue this way.

**Comment by Cryoris** [comment]
@shifubear We're moving away from opflow, and you can instead use the `PauliEvolutionGate` to construct evolutions of Pauli operators. For example:
```python
from qiskit.circuit.library import PauliEvolutionGate
from qiskit.quantum_info import SparsePauliOp

y4_hamiltonian = SparsePauliOp(["IIII", "YYYY"])  # your Hamiltonian

# this is a gate you could append to a circuit, e.g. like
# my_circuit = QuantumCircuit(4)
# my_circuit.append(evo_gate, [0, 1, 2, 3])
evo_gate = PauliEvolutionGate(y4_hamiltonian) 

# or we can just print the gate -- call .decompose() before .draw() 
# to see what's going on inside the blocks
print(evo_gate.definition.draw())
```

**Comment by 1ucian0** [comment]
With Qiskit 0.46 I was unable to reproduce this issue. The instructions to reproduce the problem result in no exception at this point and the output for `exp_ham` is:
```
global phase: 5.2832
     ┌───────────────────┐
q_0: ┤0                  ├
     │                   │
q_1: ┤1                  ├
     │  exp(-it YYYY)(1) │
q_2: ┤2                  ├
     │                   │
q_3: ┤3                  ├
     └───────────────────┘
```

The issue was probably fix between `qiskit 0.37` (which included terra 0.21) and `qiskit 0.46`.

@shifubear , let us know if updating does not solved the issue and we can reopen it.
